### PR TITLE
bpo-29636: improve CLI of `json.tool`

### PR DESCRIPTION
**Pull request by wimglenn**
Added ability to pass through `indent` and `ensure_ascii` options in `json.tool` command-line interface.

Related issues are [27413](https://bugs.python.org/issue27413) and [29636](https://bugs.python.org/issue29636).

There is other PR existing: https://github.com/python/cpython/pull/345 - that one is stagnating because @serhiy-storchaka commented on the bpo thinking it should not be merged. Serhiy claimed it adds too many (potentially conflicting) options, and I would tend to agree. This alternate implementation is a more straight-forward pass-thru of the relevant options to `json.dumps`.

Some motivation: [How to use json.tool from the shell to validate and pretty-print language files without removing the unicode?](https://stackoverflow.com/q/52712674/674039)

https://bugs.python.org/issue29636

### CODE REVIEW

File: Lib/json/tool.py
Comment (4383) [inline Lib/json/tool.py:42]: if you use type in argparse you can remove this section

**Review by 4383 — CHANGES_REQUESTED** [review]

File: Lib/json/tool.py
Comment (4383) [inline Lib/json/tool.py:27]: declare the type here for automatic cast:
```python
type=int
```

**Review by 4383 — CHANGES_REQUESTED** [review]

File: Lib/json/tool.py
Comment (wimglenn) [reply Lib/json/tool.py:42]: No, then it will be impossible to pass tab character through to json dumps from the command-line, for example by using `--indent="	"`.

File: Lib/json/tool.py
Comment (wimglenn) [reply Lib/json/tool.py:27]: Avoiding int type is intentional, since `json.dumps` can also accept integer number of spaces, or arbitrary user-provided string. This is a feature.

Comment (4383) [reply Lib/json/tool.py:27]: ack! I take notes

File: Lib/json/tool.py
Comment (4383) [reply Lib/json/tool.py:42]: ack!

**Review by 4383 — APPROVED** [review]

**Comment by dhimmel** [comment]
Author of [bpo-29636](https://bugs.python.org/issue29636) and the associated PR https://github.com/python/cpython/pull/345 to add indentation options to `json.tool`. I also have an open PR (meant to come after the `--indent` changes) in https://github.com/python/cpython/pull/201 to add `--no-ensure-ascii`. Happy to see interest from the participants here to get these features added, but I've been maintaining these open PRs for over a year now and they have undergone considerable review and discussion. Hence, I'm not sure the right way forward is for new PRs to propose the same functionality... rather than say getting involved in the BPO discussion such that consensus can be reached that these are valuable features to add.

> There is other PR existing: #345 - that one is stagnating because @serhiy-storchaka commented on the bpo thinking it should not be merged. Serhiy claimed it adds too many (potentially conflicting) options, and I would tend to agree. This alternate implementation is a more straight-forward pass-thru of the relevant options to json.dumps.

I initially started out my PR with only supporting `--indent` very similar to what is proposed here. However, discussion at https://github.com/python/cpython/pull/201#discussion_r102146742, https://github.com/python/cpython/pull/2720#discussion_r127620912, and https://github.com/python/cpython/pull/345#discussion_r103987472 pushed my PR to support more indentation options along the lines of the `jq` utility. However, as I note in the BPO discussion, I'm happy to remove any of the mutually exclusive indentation options at the maintainers / community's discretion. I just thought it'd be easiest to work backwards.

Anyways, @wimglenn happy to have you involved, if you'd like to help with any of the open PRs and discussion on BPO.

**Comment by dhimmel** [comment]
@wimglenn https://github.com/python/cpython/pull/345 was merged into https://github.com/python/cpython/commit/03257949bc02a4afdf2ea1eb07a73f8128129579 providing indentation control for json.tool.

I also have https://github.com/python/cpython/pull/201 open providing an option for ensure_ascii. I'd be happy to close that PR in favor of this PR, if you are interesting in updating the scope to just be the ascii control and maintaining the PR until its merged.

**Comment by wimglenn** [comment]
Hey Daniel, thanks for the ping. That's a little bizarre, the last discussion on [issue29636](https://bugs.python.org/issue29636) was 

@serhiy-storchaka 
> I don't think this PR should be merged. It adds too much options.

@rhettinger  
> I'm inclined to agree with Serhiy.

Then suddenly 4 days ago @methane asks "would you resolve the conflict?" and it's merged without further ado?  What changed? Did these two core developers change their mind in some offline discussion?

I mean I'm glad to have the feature in, even if I think the #345 was a little too complicated it's better than not having the functionality at all. I don't have much interest in reworking this PR for the --ensure-ascii, it's still labelled "awaiting core review" with not a peep from core in more than a year so I think I'll just be wasting my time and effort.

**Comment by methane** [comment]
I merged #345 because I want to reduce open PRs and issues.  I thought about I should reject it or merge it.
I decided to merge it because rejecting it doesn't make sense after we merged `--json-lines` option.

json-lines is not valid JSON, and output of the option is not a valid json-lines too.
I don't think the number of users of `--indent` option is significantly smaller than `--json-lines`.

**Comment by wimglenn** [comment]
@methane  "To reduce open PRs and issues" is a noble goal, but you don't request [documentation](https://docs.python.org/3/library/json.html#command-line-options) for the change to go in alongside the change itself? And to disregard/ignore the comments from other core developers?

I'll put the `--no-ensure-ascii` part to a new PR.

**Comment by methane** [comment]
> but you don't request [documentation](https://docs.python.org/3/library/json.html#command-line-options) for the change to go in alongside the change itself?

Yes, I noticed I missed it.  It should be added.

> And to disregard/ignore the comments from other core developers?

Sometime, it is difficult to make a final agreement.

We discuss in python-dev mailing list If the problem is important enough.
But in this case, I didn't think this issue is important enough.  So I merged it by my own judgment.

I think this is important to reduce open PRs and issues.  We can not discuss and make a decision in the mailing list for all feature requests.  Time is a limited resource.

Of course, I might be wrong and other core devs oppose it strongly.  We can revert it then.

**Comment by wimglenn** [comment]
Here is the other part --> https://github.com/python/cpython/pull/17472

**Comment by dhimmel** [comment]
> Why was it merged with no documentation, there should be a change in `Doc/library/json.rst` "Command line options" section?

I am happy to make another PR to add this documentation for #345.

**Pull request closed without merging**
